src/haplotypes/model.rs

```diff
@@ -124,7 +126,41 @@ impl Likelihood {
     ) -> LogProb {
         // TODO compute likelihood based on Varlociraptor VAFs.
         // Let us postpone this until we have a working version with kallisto only.
-        LogProb::ln_one()
+        let variant_matrix: Vec<BitVec> = data.haplotype_variants.values().cloned().collect();
+        let likelihood = Prob(1.0);
+        for (variant, bv) in data.haplotype_variants.iter() {
+            let variant_sum = 0.0;
+            for (i, fraction) in event.iter().enumerate() {
+                if bv[i.try_into().unwrap()] == true {
+                    let variant_sum = variant_sum + NotNaN::into_inner(*fraction);
+                }
+            }
+            let mut vafs = Vec::new();
+            let mut densities = Vec::new();
+            if data.haplotype_calls.contains_key(variant) {
+                let afd = data.haplotype_calls[variant].split(",");
+                for pair in afd {
+                    let (vaf, density) = pair.split_once("=").unwrap();
+                    let (vaf, density): (f64, f64) =
+                        (vaf.parse().unwrap(), density.parse().unwrap());
+                    vafs.push(vaf);
+                    densities.push(density);
+                }
+                for (i, (vaf, density)) in vafs.iter().zip(densities.iter()).enumerate() {
+                    let vaf_sum = f64::from(Prob::from(PHREDProb(variant_sum)));
+                    if vaf == &vaf_sum {
+                        let likelihood = likelihood * Prob::from(PHREDProb(*density));
+                    } else {
+                        let density = densities[i - 1]
+                            + (vaf_sum - vafs[i - 1]) * (densities[i + 1] - densities[i - 1])
+                                / (vafs[i + 1] - vafs[i - 1]);
+                        let likelihood = likelihood * Prob::from(PHREDProb(density));
+                    }
+                }
```

Comment:
This goes in the right direction, but is not ideal yet because it is slow. I have above suggested to use a HashMap for storing the densities. Let me revise that, better use a BTreeMap. Then, you can actually get either the exact matching entry or those two surrounding your desired VAF at logarithmic time, with API calls instead of the need to write a loop here. And the code for that kind of lookup should be put as a method into the HaplotypeCalls struct (the paradigm that applies here is "separation of concerns").